File: LeetCode/Problem 234.py
# LeetCode: Problem 234
# Jonathan Kosasih

#Given a singly linked list, determine if it is a palindrome.

#Example 1:
#Input: 1->2
#Output: false
#
#Example 2:
#Input: 1->2->2->1
#Output: true
#------------------------------------------------------------------
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    # O(n) time, O(n) space
    def isPalindrome(self, head: ListNode) -> bool:
        nodes = []
        while(head):
            nodes.append(head.val)
            head = head.next
        return nodes == nodes[::-1]
    
    # O(n) time, O(1) space
    def isPalindromeAlt(self, head: ListNode) -> bool:
        slow = fast = head
        # find midpoint
        while(fast and fast.next):
            fast = fast.next.next
            slow = slow.next
        node = None
        # found midpoint, now reverse the 2nd half
        while(slow):
            temp = slow.next
            slow.next = node
            node = slow
            slow = temp
        # compare from beginning to reversed 2nd half
        while(node):
            if(node.val != head.val):
                return False
            node = node.next
            head = head.next
        return True
            
sol = Solution()
node1 = ListNode(1)
node2 = ListNode(2)
node3 = ListNode(2)
node4 = ListNode(1)
node1.next = node2
node2.next = node3
node3.next = node4
print(str(sol.isPalindrome(node1)))
node1 = ListNode(-129)
node2 = ListNode(-129)
node1.next = node2
print(str(sol.isPalindrome(node1)))

File: LeetCode/C++/Problem 234.cpp
// LeetCode: Problem 234
// Jonathan Kosasih
/*
Given a singly linked list, determine if it is a palindrome.
*/
//-------------------------------------------------------------------

#include <iostream>
#include <vector>
#include <map>
#include <unordered_map>
#include <set>
#include <string>
#include <algorithm> 
#include <stack>
#include <queue>
#include <deque>
#include <numeric>
using std::vector;
using std::map;
using std::unordered_map;
using std::set;
using std::stack;
using std::queue;
using std::deque;
using std::accumulate;

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	bool isPalindrome(ListNode* head) {
		if (head == nullptr || head->next == nullptr) return true;

		// Find middle & reverse 2nd half of the list
		ListNode* reversed2ndHalf = reverseList(findMiddle(head));

		// Verify palindrome property
		while (reversed2ndHalf != nullptr && head)
		{
			if (reversed2ndHalf->val != head->val)
				return false;

			reversed2ndHalf = reversed2ndHalf->next;
			head = head->next;
		}

		return true;
	}

private:
	ListNode* findMiddle(ListNode* head)
	{
		if (head == nullptr || head->next == nullptr) return head;

		ListNode* slow = head; ListNode* fast = head;
		while (fast != nullptr && fast->next != nullptr)
		{
			slow = slow->next;
			fast = fast->next->next;
		}

		// If we have an odd number of elements skip over the middle element
		if (fast != nullptr) slow = slow->next;

		return slow;
	}

	ListNode* reverseList(ListNode* head)
	{
		if (head == nullptr || head->next == nullptr) return head;

		ListNode* prev = nullptr;
		while (head != nullptr)
		{
			ListNode* next = head->next;
			head->next = prev;
			prev = head;
			head = next;
		}

		return prev;
	}
};
int main()
{
    //std::cout << "Hello World!\n";
}